src/magic_spell_effect.cpp

```diff
@@ -481,6 +482,69 @@ void spell_effect::target_attack( const spell &sp, Creature &caster,
     }
 }
 
+static void magical_polymorph( monster &victim, Creature &caster, const spell &sp )
+{
+    std::string new_id = sp.effect_data();
+
+    if( sp.has_flag( spell_flag::POLYMORPH_GROUP ) ) {
+        const mongroup_id group_id( sp.effect_data() );
+        new_id = MonsterGroupManager::GetRandomMonsterFromGroup( group_id ).str();
+    }
+
+    // if effect_str is empty, we become a random monster of close difficulty
+    if( new_id.empty() ) {
+        int victim_diff = victim.type->difficulty;
+        std::vector<mtype> mtypes = MonsterGenerator::generator().get_all_mtypes();
```

Comment:
```suggestion
        const std::vector<mtype> &mtypes = MonsterGenerator::generator().get_all_mtypes();
```
No need to copy all of them.